File: api/mcfrs_api/services/privacy.py
"""Privacy transforms, applied SERVER-SIDE before an incident reaches a public
browser. Member mode passes everything through. Public mode strips what could
identify a patient or a residence for medical calls."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

from mcfrs_api.models import Incident, IncidentCategory

MEDICAL = {IncidentCategory.EMS, IncidentCategory.ALS, IncidentCategory.BLS, IncidentCategory.MASS_CASUALTY}
_HOUSE = re.compile(r"^\s*(\d+)([A-Za-z]?)\s+(.+)$")

Precision = Literal["exact", "block", "street"]


@dataclass(frozen=True)
class PrivacyOptions:
    mode: Literal["member", "public"] = "member"
    mask_medical_addresses: bool = True
    precision: Precision = "block"


def is_medical(category: IncidentCategory) -> bool:
    return category in MEDICAL


def mask_address(address: str, precision: Precision) -> str:
    """'12345 Example Farm Rd' -> block: '12300 blk Example Farm Rd', street: 'Example Farm Rd'."""
    if precision == "exact":
        return address
    m = _HOUSE.match(address)
    if not m:
        return address  # no house number: nothing to hide
    num, _, street = m.groups()
    if precision == "street":
        return street
    return f"{int(num) // 100 * 100} blk {street}"


def fuzz_coordinate(v: float, step: float = 0.002) -> float:
    """Snap to ~0.002 deg (~200 m) so a sensitive point is not a rooftop."""
    return round(round(v / step) * step, 6)


def apply_privacy(incident: Incident, opts: PrivacyOptions) -> Incident:
    if opts.mode == "member":
        return incident
    out = incident.model_copy(update={"raw_payload": None})
    if opts.mask_medical_addresses and is_medical(incident.category):
        update: dict[str, object] = {"cross_street": None}
        if out.address:
            update["address"] = mask_address(out.address, opts.precision)
        if out.latitude is not None:
            update["latitude"] = fuzz_coordinate(out.latitude)
        if out.longitude is not None:
            update["longitude"] = fuzz_coordinate(out.longitude)
        out = out.model_copy(update=update)
    return out

File: src/lib/privacy.ts
import type { Incident, IncidentCategory } from "@/types/incident";

/**
 * Privacy transforms, applied SERVER-SIDE before an incident reaches a public browser.
 * Member mode passes everything through. Public mode strips what could identify a
 * patient or a residence for medical calls.
 */
export interface PrivacyOptions {
  mode: "member" | "public";
  maskMedicalAddresses: boolean;
  precision: "exact" | "block" | "street";
}

const MEDICAL: ReadonlySet<IncidentCategory> = new Set(["ems", "als", "bls", "mass_casualty"]);

export function isMedical(category: IncidentCategory): boolean {
  return MEDICAL.has(category);
}

/** "12345 Example Farm Rd" -> block: "12300 blk Example Farm Rd", street: "Example Farm Rd". */
export function maskAddress(address: string, precision: "exact" | "block" | "street"): string {
  if (precision === "exact") return address;
  const m = address.match(/^\s*(\d+)([A-Za-z]?)\s+(.+)$/);
  if (!m) return address; // no house number: nothing to hide
  const [, num, , street] = m;
  if (precision === "street") return street;
  const n = parseInt(num, 10);
  const block = Math.floor(n / 100) * 100;
  return `${block} blk ${street}`;
}

/** Snap a coordinate to ~0.002° (~200 m) so a sensitive point is not a rooftop. */
export function fuzzCoordinate(v: number, step = 0.002): number {
  return Math.round(v / step) * step;
}

export function applyPrivacy(incident: Incident, opts: PrivacyOptions): Incident {
  if (opts.mode === "member") return incident;
  const sensitive = opts.maskMedicalAddresses && isMedical(incident.category);
  const out: Incident = { ...incident, rawPayload: undefined };
  if (sensitive) {
    if (out.address) out.address = maskAddress(out.address, opts.precision);
    out.crossStreet = undefined;
    if (typeof out.latitude === "number") out.latitude = fuzzCoordinate(out.latitude);
    if (typeof out.longitude === "number") out.longitude = fuzzCoordinate(out.longitude);
  }
  return out;
}
